File: jbake-core/src/test/java/org/jbake/launcher/LaunchOptionsTest.java
package org.jbake.launcher;

import org.jbake.app.configuration.ConfigUtil;
import org.junit.Test;
import picocli.CommandLine;
import picocli.CommandLine.MissingParameterException;

import java.io.File;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatExceptionOfType;

public class LaunchOptionsTest {

    @Test
    public void showHelp() {
        String[] args = {"-h"};
        LaunchOptions res = parseArgs(args);
        assertThat(res.isHelpNeeded()).isTrue();
    }

    @Test
    public void runServer() {
        String[] args = {"-s"};
        LaunchOptions res = parseArgs(args);

        assertThat(res.isRunServer()).isTrue();
    }

    @Test
    public void runServerWithFolder() {
        String[] args = {"-s", "/tmp"};
        LaunchOptions res = parseArgs(args);

        assertThat(res.isRunServer()).isTrue();
        assertThat(res.getSource()).isEqualTo(new File("/tmp"));
    }

    @Test
    public void init() {
        String[] args = {"-i"};
        LaunchOptions res = parseArgs(args);

        assertThat(res.isInit()).isTrue();
        assertThat(res.getTemplate()).isEqualTo("freemarker");
    }

    @Test
    public void initWithTemplate() {
        String[] args = {"-i", "-t", "foo"};
        LaunchOptions res = parseArgs(args);

        assertThat(res.isInit()).isTrue();
        assertThat(res.getTemplate()).isEqualTo("foo");
    }

    @Test
    public void initWithSourceDirectory() {
        String[] args = {"-i", "/tmp"};
        LaunchOptions res = parseArgs(args);

        assertThat(res.isInit()).isTrue();
        assertThat(res.getSourceValue()).isEqualTo("/tmp");
    }

    @Test
    public void initWithTemplateAndSourceDirectory() {
        String[] args = {"-i", "-t", "foo", "/tmp"};
        LaunchOptions res = parseArgs(args);

        assertThat(res.isInit()).isTrue();
        assertThat(res.getTemplate()).isEqualTo("foo");
        assertThat(res.getSourceValue()).isEqualTo("/tmp");
    }

    @Test
    public void shouldThrowAnExceptionCallingTemplateWithoutInitOption() {
        String[] args = {"-t", "groovy-mte"};

        assertThatExceptionOfType(MissingParameterException.class).isThrownBy(()-> {
            LaunchOptions res = parseArgs(args);
        }).withMessage("Error: Missing required argument(s): --init");
    }

    @Test
    public void bake() {
        String[] args = {"-b"};
        LaunchOptions res = parseArgs(args);

        assertThat(res.isBake()).isTrue();
    }

    @Test
    public void listConfig() throws Exception {
        String[] args = {"-ls"};
        LaunchOptions res = parseArgs(args);

        assertThat(res.isListConfig()).isTrue();
    }

    @Test
    public void listConfigLongOption() throws Exception {
        String[] args = {"--list-settings"};
        LaunchOptions res = parseArgs(args);

        assertThat(res.isListConfig()).isTrue();
    }

    @Test
    public void customPropertiesEncoding() throws Exception {
        String[] args = {"--prop-encoding", "utf-16"};
        LaunchOptions res = parseArgs(args);

        assertThat(res.getPropertiesEncoding()).isEqualTo("utf-16");
    }

    @Test
    public void defaultEncodingIsUtf8() throws Exception {
        String[] args = {};
        LaunchOptions res = parseArgs(args);

        assertThat(res.getPropertiesEncoding()).isEqualTo("utf-8");
    }

    @Test
    public void bakeNoArgs() {
        String[] args = {};
        LaunchOptions res = parseArgs(args);

        assertThat(res.isHelpNeeded()).isTrue();
        assertThat(res.isRunServer()).isFalse();
        assertThat(res.isInit()).isFalse();
        assertThat(res.isBake()).isFalse();
        assertThat(res.getSource().getPath()).isEqualTo(System.getProperty("user.dir"));
        assertThat(res.getDestination().getPath()).isEqualTo(System.getProperty("user.dir") + File.separator + "output");
        assertThat(res.getConfig().getPath()).isEqualTo(System.getProperty("user.dir") + File.separator + ConfigUtil.CONFIG_FILE);
    }

    @Test
    public void bakeWithArgs() {
        String[] args = {"/tmp/source", "/tmp/destination"};
        LaunchOptions res = parseArgs(args);

        assertThat(res.isHelpNeeded()).isFalse();
        assertThat(res.isRunServer()).isFalse();
        assertThat(res.isInit()).isFalse();
        assertThat(res.isBake()).isTrue();
        assertThat(res.getSource()).isEqualTo(new File("/tmp/source"));
        assertThat(res.getDestination()).isEqualTo(new File("/tmp/destination"));
    }

    @Test
    public void configArg() {
        String[] args = {"-c", "foo"};
        LaunchOptions res = parseArgs(args);
        assertThat(res.getConfig().getAbsoluteFile().toString()).isEqualTo(System.getProperty("user.dir")+ File.separator + "foo");

    }

    private LaunchOptions parseArgs(String[] args) {
        return CommandLine.populateCommand(new LaunchOptions(), args);
    }
}

File: jbake-core/src/test/java/org/jbake/launcher/LaunchOptionsTest.kt
package org.jbake.launcher

import io.kotest.assertions.throwables.shouldThrow
import io.kotest.core.spec.style.StringSpec
import io.kotest.matchers.shouldBe
import org.jbake.app.configuration.ConfigUtil
import org.jbake.util.PathUtils.SYSPROP_USER_DIR
import picocli.CommandLine
import java.io.File

class LaunchOptionsTest : StringSpec({

    fun parseArgs(args: Array<String>): LaunchOptions {
        return CommandLine.populateCommand(LaunchOptions(), *args)
    }

    "showHelp" {
        val args = arrayOf("-h")
        val res = parseArgs(args)
        res.isHelpNeeded shouldBe true
    }

    "runServer" {
        val args = arrayOf("-s")
        val res = parseArgs(args)

        res.isRunServer shouldBe true
    }

    "runServerWithFolder" {
        val args = arrayOf("-s", "/tmp")
        val res = parseArgs(args)

        res.isRunServer shouldBe true
        res.getSource() shouldBe File("/tmp")
    }

    "init" {
        val args = arrayOf("-i")
        val res = parseArgs(args)

        res.isInit shouldBe true
        res.template shouldBe "freemarker"
    }

    "initWithTemplate" {
        val args = arrayOf("-i", "-t", "foo")
        val res = parseArgs(args)

        res.isInit shouldBe true
        res.template shouldBe "foo"
    }

    "initWithSourceDirectory" {
        val args = arrayOf("-i", "/tmp")
        val res = parseArgs(args)

        res.isInit shouldBe true
        res.sourceValue shouldBe "/tmp"
    }

    "initWithTemplateAndSourceDirectory" {
        val args = arrayOf("-i", "-t", "foo", "/tmp")
        val res = parseArgs(args)

        res.isInit shouldBe true
        res.template shouldBe "foo"
        res.sourceValue shouldBe "/tmp"
    }

    "shouldThrowAnExceptionCallingTemplateWithoutInitOption" {
        val args = arrayOf("-t", "groovy-mte")

        shouldThrow<CommandLine.MissingParameterException> {
            parseArgs(args)
        }
    }

    "bake" {
        val args = arrayOf("-b")
        val res = parseArgs(args)

        res.isBake shouldBe true
    }

    "listConfig" {
        val args = arrayOf("-ls")
        val res = parseArgs(args)

        res.isListConfig shouldBe true
    }

    "listConfigLongOption" {
        val args = arrayOf("--list-settings")
        val res = parseArgs(args)

        res.isListConfig shouldBe true
    }

    "customPropertiesEncoding" {
        val args = arrayOf("--prop-encoding", "utf-16")
        val res = parseArgs(args)

        res.propertiesEncoding shouldBe "utf-16"
    }

    "defaultEncodingIsUtf8" {
        val args = arrayOf<String>()
        val res = parseArgs(args)

        res.propertiesEncoding shouldBe "utf-8"
    }

    "bakeNoArgs" {
        val args = arrayOf<String>()
        val res = parseArgs(args)

        res.isHelpNeeded shouldBe true
        res.isRunServer shouldBe false
        res.isInit shouldBe false
        res.isBake shouldBe false
        res.getSource().path shouldBe SYSPROP_USER_DIR
        res.getDestination().path shouldBe File(SYSPROP_USER_DIR, "output").path
        res.getConfig().path shouldBe File(SYSPROP_USER_DIR, ConfigUtil.CONFIG_FILE).path
    }

    "bakeWithArgs" {
        val args = arrayOf("/tmp/source", "/tmp/destination")
        val res = parseArgs(args)

        res.isHelpNeeded shouldBe false
        res.isRunServer shouldBe false
        res.isInit shouldBe false
        res.isBake shouldBe true
        res.getSource() shouldBe File("/tmp/source")
        res.getDestination() shouldBe File("/tmp/destination")
    }

    "configArg" {
        val args = arrayOf("-c", "foo")
        val res = parseArgs(args)
        res.getConfig().absoluteFile.toString() shouldBe File(SYSPROP_USER_DIR, "foo").absoluteFile.toString()
    }
})
